### PR TITLE
Initial mesh tools branch for TensorMeshes

### PR DESCRIPTION
**Pull request by prisae**
# Add more meshing tools to create fancy meshes

## Motivation
We probably all have our own little routines to create more complex meshes. I think it would be great to incorporate some of these into discretize, so not everyone has to do the same again.

## Status so far
Here, I included my most used routines:
- `get_domain`: A routine that calculates minimum cell width and calculation extent based on multiples of the skin depth. The user can restrict the domain to a minimum domain, and give a range for the possible min-cell width. Good for on-the-fly mesh creations.
- `get_stretched_h`: Get the input (together with `get_domain`) for a `TensorMesh` for a certain domain size, a certain number of cells, a minimum cell width, and an optional domain without stretching. The routine finds the suitable stretching factor.

## Warning
- This is an initial draft. I tried to document as good as possible, so it should be easy to incorporate other ideas and stuff from other routines. Please add your own stuff!
- I added tests for 100% coverage. However, in pytest, so they probably will fail hilariously. I thought I won't bother checking if they are good for unittest as long as it is in an early stage.
- I also added an initial notebook, didn't check, will probably fail too.

### CODE REVIEW

**Comment by prisae** [comment]
Files changed:
- `discretize/utils/discretize/meshutils.py` => Routines
- `docs/exapmles/plot_meshutils.py` => Example
- `tests/base/test_utils.py` => Tests

Let me know if they would go better somewhere else.

**Comment by prisae** [comment]
Here is the notebook, which I also added as an example:

https://github.com/prisae/tmp-share/blob/master/TensorMesh-utilities.ipynb

File: docs/examples/plot_meshutils.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+"""
```
Comment (rowanc1) [inline docs/examples/plot_meshutils.py]: Thanks for adding the example!

**Comment by thast** [comment]
Just to mention it, @sgkang and @lheagy had a utils to create 1D mesh for MT in their Leading Edge MT tutorial. Maybe worth a look too: https://github.com/simpeg/tle-magnetotelluric_inversion/blob/49b038b013a40283a32ea18ea9144bf707656873/tleMT/MT1D.py#L72

**Comment by prisae** [comment]
Thanks @thast, exactly things like that could be added. I personally won't add anything more, the ones I added are the ones I use in my current project, but I invite everyone to add more tools. Although this could also be done at a later stage, of course.

I meant this PR also to discuss the topic in general. Is `meshutils` the right place for it? Or should we start a dedicated file just for such routines? Any input/idea/suggestion around the topic of mesh generation is welcomed here. 

My meshing routine is only good in some circumstances. For instance, if you have a fixed number of cells you want to use (like I use only powers of two, 2^x, currently). If you need large padding, like in my case, I set the air-padding to something like 40*skindepth. But obviously, for other applications these settings won't suit.

**Comment by fourndo** [comment]
That's great @prisae . I worked on something similar, mostly targeting for 3D TreeMesh on #141 
Called it ```meshBuilderXYZ``` right now.
Maybe we should see if we can mash-up the two ...

**Comment by prisae** [comment]
@fourndo that would be great and exactly what I am looking for. To combine our many approaches and create a few widely applicable helper routines.

**Comment by prisae** [comment]
Had a look at your function. So the main difference is:
- Your function uses a fixed stretching factor, and finds the required number of cells.
- My function uses a fixed number of cells, and tries to find the stretching factor. 

On top of that, you have the whole Tree-Mesh stuff.

An underlying design question is what it returns. You return a 3D mesh. In my case, it just returns the cell widths, you still have to call TensorMesh with it. The downside of my approach is that more lines of codes are needed when you use it, but the upside is that it keeps it much more general, as it can be used for 1D, 2D, and 3D.
```
hx = discretize.utils.meshutils.get_stretched_h(hx_min, xdomain, nx, x0, x1)
hy = discretize.utils.meshutils.get_stretched_h(hy_min, ydomain, ny, y0, y1)
hz = discretize.utils.meshutils.get_stretched_h(hz_min, zdomain, nz, z0, z1)
mesh = discretize.TensorMesh([hx, hy, hz], x0=[xdomain[0], ydomain[0], zdomain[0]])
```

I think if we want to combine them the first step would be to split up yours into a similar scheme, so it only returns cell widths, with which you can call TensorMesh. If this is something you don't want to consider, then we keep them better as separate options.

**Comment by prisae** [comment]
I am closing this. It is sort of outdated, I think everybody moved on and uses some functionalities etc.

It would still be great if we have some automated meshing functionalities, but I think we would have to start fresh.

**Pull request closed without merging**